src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,138 @@
+import random
+from typing import List, Optional
+
+import torch
+
+from torchcodec.decoders import FrameBatch, SimpleVideoDecoder
+
+
+def _validate_params(
+    *, decoder, num_clips, num_frames_per_clip, num_indices_between_frames
+):
+    if len(decoder) < 1:
+        raise ValueError(
+            f"Decoder must have at least one frame, found {len(decoder)} frames."
+        )
+
+    if num_clips <= 0:
+        raise ValueError(f"num_clips ({num_clips}) must be strictly positive")
+    if num_frames_per_clip <= 0:
+        raise ValueError(
+            f"num_frames_per_clip ({num_frames_per_clip}) must be strictly positive"
+        )
+    if num_indices_between_frames <= 0:
+        raise ValueError(
+            f"num_indices_between_frames ({num_indices_between_frames}) must be strictly positive"
+        )
+
+
+def _validate_sampling_range(
+    *, sampling_range_start, sampling_range_end, num_frames, clip_span
+):
+    if sampling_range_start < 0:
+        sampling_range_start = num_frames + sampling_range_start
```

Comment:
what if it's still negative after this?

Reply:
It would result in an error because `sampling_range_end <= sampling_range_start` .

If we don't error, we can either:

- decide that it's equivalent to 1 (so that it can work when sampling_range_start is 0)
- keep wrapping-around (modulo).

Neither sounds like a great intuitive option, so I it might be best to error, similar to what we decided in https://github.com/pytorch/torchcodec/pull/221/files#r1781347650